stdlib/public/core/Substring.swift

```diff
@@ -117,104 +126,374 @@ public struct Substring: Sendable {
+    let nextStride = _characterStride(startingAt: nextIndex)
+
+    var r = Index(
+      encodedOffset: nextOffset, characterStride: nextStride)._scalarAligned
+
+    // Don't set the `_isCharacterAligned` bit in indices of exotic substrings
+    // whose startIndex isn't aligned on a grapheme cluster boundary. (Their
+    // grapheme breaks may not match with those in `base`.)
+    //
+    // Note that we don't need to care about whether the end index is aligned
+    // here.
+    if _startIsCharacterAligned {
+      r = r._characterAligned
+    }
+
+    return r
   }
 
-  @inlinable @inline(__always)
   public func index(before i: Index) -> Index {
-    _precondition(i <= endIndex, "Cannot decrement an invalid index")
-    _precondition(i > startIndex, "Cannot decrement beyond startIndex")
-    return _slice.index(before: i)
+    // Note: in Swift 5.6 and below, this method used to be inlinable,
+    // forwarding to `_slice.base.index(before:)`. Unfortunately, that approach
+    // isn't compatible with SE-0180, as it allows Unicode scalars outside the
+    // substring to affect grapheme breaking results within the substring. This
+    // leads to Collection conformance issues when the `Substring`'s bounds do
+    // not fall on grapheme boundaries in `base`.
+
+    let i = _roundDownToNearestCharacter(_validateInclusiveScalarIndex(i))
+    // Note: Aligning an index may move it closer towards the `startIndex`, so
+    // this `i > startIndex` check needs to come after all the
+    // alignment/validation work.
+    _precondition(i > startIndex, "Substring index is out of bounds")
+
+    let r = _uncheckedIndex(before: i)
+    return _wholeGuts.markEncoding(r)
   }
 
-  @inlinable @inline(__always)
-  public func index(_ i: Index, offsetBy n: Int) -> Index {
-    let result = _slice.index(i, offsetBy: n)
-    _precondition(
-      (_slice._startIndex ... _slice.endIndex).contains(result),
-      "Operation results in an invalid index")
-    return result
+  /// A version of `index(before:)` that assumes that the given index:
+  ///
+  /// - has the right encoding,
+  /// - is within bounds, and
+  /// - is character aligned within this substring.
+  ///
+  /// It does not mark the encoding of the returned index.
+  internal func _uncheckedIndex(before i: Index) -> Index {
+    _internalInvariant(_wholeGuts.hasMatchingEncoding(i))
+    _internalInvariant(i._isScalarAligned)
+    _internalInvariant(i > startIndex && i <= endIndex)
+
+    // Note: `i` must be `Character`-aligned within this substring, even if it
+    // doesn't have the corresponding flag set.
+
+    // TODO: known-ASCII fast path, single-scalar-grapheme fast path, etc.
+    let priorStride = _characterStride(endingAt: i)
+    let priorOffset = i._encodedOffset &- priorStride
+    _internalInvariant(priorOffset >= startIndex._encodedOffset)
+
+    var r = Index(
+      encodedOffset: priorOffset, characterStride: priorStride)
```

Comment:
```suggestion
    var r = Index(
      encodedOffset: priorOffset,
      characterStride: priorStride
    )._scalarAligned
```

so you can remove the else branch below